openmc/cmfd.py

```diff
@@ -0,0 +1,3072 @@
+        matrix : scipy.sparse.spmatrix
+            Sparse matrix that will be saved
+        base_filename : str
+            Filename to save matrix entries, without any file extension at the
+            end. Matrix entries will be saved to file [base_filename].dat and
+            in scipy format as [base_filename].npz
+
+        """
+        # Write row, col, and data of each entry in sparse matrix. This ignores
+        # all zero-entries, and indices are written with zero-based indexing
+        with open(base_filename+'.dat', 'w') as fh:
+            for row in range(matrix.shape[0]):
+                # Get all cols for particular row in matrix
+                cols = matrix.indices[matrix.indptr[row]:matrix.indptr[row+1]]
+                # Get all data entries for particular row in matrix
+                data = matrix.data[matrix.indptr[row]:matrix.indptr[row+1]]
+                for i in range(len(cols)):
+                    fh.write('{:3d}, {:3d}, {:0.8f}\n'.format(
+                        row, cols[i], data[i]))
+
+        # Save matrix in scipy format
+        sparse.save_npz(base_filename, matrix)
+
+    def _calc_fission_source(self):
+        """Calculates CMFD fission source from CMFD flux. If a coremap is
+        defined, there will be a discrepancy between the spatial indices in the
+        variables ``phi`` and ``nfissxs``, so ``phi`` needs to be mapped to the
+        spatial indices of the cross sections. This can be done in a vectorized
+        numpy manner or with for loops
+
+        """
+        # Extract number of groups and number of accelerated regions
+        nx, ny, nz, ng = self._indices
+        n = self._mat_dim
+
+        # Compute cmfd_src in a vecotorized manner by phi to the spatial
+        # indices of the actual problem so that cmfd_flux can be multiplied by
+        # nfissxs
+
+        # Calculate volume
+        vol = np.product(self._hxyz, axis=3)
+
+        # Reshape phi by number of groups
+        phi = self._phi.reshape((n, ng))
+
+        # Extract indices of coremap that are accelerated
+        idx = self._accel_idxs
+
+        # Initialize CMFD flux map that maps phi to actual spatial and
+        # group indices of problem
+        cmfd_flux = np.zeros((nx, ny, nz, ng))
+
+        # Loop over all groups and set CMFD flux based on indices of
+        # coremap and values of phi
+        for g in range(ng):
+            phi_g = phi[:,g]
+            cmfd_flux[idx + (g,)] = phi_g[self._coremap[idx]]
+
+        # Compute fission source
+        cmfd_src = (np.sum(self._nfissxs[:,:,:,:,:] *
+                    cmfd_flux[:,:,:,:,np.newaxis], axis=3) *
+                    vol[:,:,:,np.newaxis])
+
+        # Normalize source such that it sums to 1.0
+        self._cmfd_src = cmfd_src / np.sum(cmfd_src)
+
+        # Compute entropy
+        if openmc.lib.settings.entropy_on:
+            # Compute source times log_2(source)
+            source = self._cmfd_src[self._cmfd_src > 0] \
```

Comment:
Excellent use of numpy advanced boolean array indexing for a filtering operation.